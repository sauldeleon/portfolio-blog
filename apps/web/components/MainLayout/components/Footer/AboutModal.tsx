import {
  FigmaIcon,
  GetTheIdeaIcon,
  GithubIcon,
  JestIcon,
  LinkedInIcon,
  NextJSIcon,
  NxIcon,
  StyledComponentsIcon,
  YarnIcon,
} from '@sdlgr/assets'
import { Body } from '@sdlgr/typography'

import { useClientTranslation } from '@web/i18n/client'

import {
  StyledIconLink,
  StyledLabel,
  StyledList,
  StyledModal,
  StyledModalHeading,
  StyledPropertyWrapper,
} from './AboutModal.styles'

interface AboutModalProps {
  isOpen: boolean
  setIsOpen: React.Dispatch<React.SetStateAction<boolean>>
}

type IconListItemProps = {
  key: string
  href: string
  ariaLabel: string
  icon: React.ReactNode
}

export function AboutModal({ isOpen, setIsOpen }: AboutModalProps) {
  const { t } = useClientTranslation('footer')
  const developmentIcons: IconListItemProps[] = [
    {
      key: 'nextjs',
      href: 'https://nextjs.org/',
      ariaLabel: 'NextJS',
      icon: <NextJSIcon width={20} height={20} />,
    },
    {
      key: 'styled-components',
      href: 'https://styled-components.com/',
      ariaLabel: 'Styled-components',
      icon: <StyledComponentsIcon width={20} height={20} />,
    },
    {
      key: 'jest',
      href: 'https://jestjs.io/',
      ariaLabel: 'Jest',
      icon: <JestIcon width={20} height={20} />,
    },
    {
      key: 'nx',
      href: 'https://nx.dev/',
      ariaLabel: 'NX',
      icon: <NxIcon width={20} height={20} />,
    },
    {
      key: 'yarn',
      href: 'https://yarnpkg.com/',
      ariaLabel: 'Yarn',
      icon: <YarnIcon width={20} height={20} />,
    },
  ]

  const repositoryIcons: IconListItemProps[] = [
    {
      key: 'github',
      href: 'https://github.com/sauldeleon/portfolio-blog',
      ariaLabel: t('checkRepoHereAria'),
      icon: <GithubIcon width={20} />,
    },
  ]

  const designerIcons: IconListItemProps[] = [
    {
      key: 'linkedIn',
      href: 'https://www.linkedin.com/in/valentina-florentina-balta-cojocaru-stan-83619014a/',
      ariaLabel: t('checkLinkedInProfileAria', {
        name: 'Valentina Florentina Balta-Cojocaru-Stan',
      }),
      icon: <LinkedInIcon width={20} height={20} />,
    },
    {
      key: 'getTheIdea',
      href: 'http://valentina.pro/index.html',
      ariaLabel: t('getTheIdeaAria', {
        name: 'Valentina Florentina Balta-Cojocaru-Stan',
      }),
      icon: <GetTheIdeaIcon width={20} />,
    },
  ]

  const designIcons: IconListItemProps[] = [
    {
      key: 'figma',
      href: 'https://www.figma.com/file/ZgfaiU473XMQAWz9kbiMO6/Saul-portfolio-%2B-design-library',
      ariaLabel: t('checkDesignHereAria'),
      icon: <FigmaIcon width={20} height={20} />,
    },
  ]

  const modalSections = [
    {
      label: t('development'),
      key: 'development',
      properties: [
        { label: t('usedTechs'), key: 'usedTechs', icons: developmentIcons },
        { label: t('repository'), key: 'repository', icons: repositoryIcons },
      ],
    },
    {
      label: t('design'),
      key: 'design',
      properties: [
        { label: t('designer'), key: 'designer', icons: designerIcons },
        { label: t('designLink'), key: 'designLink', icons: designIcons },
      ],
    },
  ]

  return (
    <StyledModal isOpen={isOpen} setIsOpen={setIsOpen}>
      <section>
        <StyledModalHeading $level={2} as="h4" id="about-modal-label">
          {t('aboutModalTitle')}
        </StyledModalHeading>

        {modalSections.map(({ label, key, properties }) => (
          <div key={key}>
            <StyledLabel $level="L">{label}</StyledLabel>
            {properties.map(({ label, key, icons }) => (
              <StyledPropertyWrapper key={key}>
                <Body>{label}</Body>
                <StyledList>
                  {icons.map(({ key, href, icon, ariaLabel }) => (
                    <li key={key}>
                      <StyledIconLink aria-label={ariaLabel} href={href}>
                        {icon}
                      </StyledIconLink>
                    </li>
                  ))}
                </StyledList>
              </StyledPropertyWrapper>
            ))}
          </div>
        ))}
      </section>
    </StyledModal>
  )
}
