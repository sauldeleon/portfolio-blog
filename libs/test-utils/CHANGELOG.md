# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
