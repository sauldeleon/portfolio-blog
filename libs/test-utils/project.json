{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
