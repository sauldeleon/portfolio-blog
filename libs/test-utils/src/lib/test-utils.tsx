import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
