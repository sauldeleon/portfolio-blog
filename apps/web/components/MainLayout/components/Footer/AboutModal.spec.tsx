import { fireEvent, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderApp } from '@sdlgr/test-utils'

import { AboutModal } from './AboutModal'

describe('AboutModal', () => {
  it('should render', async () => {
    const { baseElement } = renderApp(
      <AboutModal isOpen setIsOpen={jest.fn()} />,
    )
    await screen.findByText('About this page')
    expect(baseElement).toMatchSnapshot()
  })

  it('should close', async () => {
    const mockFn = jest.fn()
    renderApp(<AboutModal isOpen setIsOpen={mockFn} />)
    const closeButton = await screen.findByText('close.svg')
    await userEvent.click(closeButton)
    expect(mockFn).toHaveBeenNthCalledWith(1, false)
  })

  it('should close on backdrop event', async () => {
    const mockFn = jest.fn()
    renderApp(<AboutModal isOpen setIsOpen={mockFn} />)
    const closeButton = await screen.findByText('close.svg')
    fireEvent.keyDown(closeButton, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27,
    })
    expect(mockFn).toHaveBeenNthCalledWith(1, false)
  })
})
