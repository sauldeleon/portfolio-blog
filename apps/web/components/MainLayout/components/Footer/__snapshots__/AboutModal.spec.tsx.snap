// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutModal should render 1`] = `
.c6 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 1.5rem;
  line-height: normal;
  margin: 0;
}

.c11 {
  font-family: 'Roboto mono',monospace;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c11 ul,
.c11 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c11 ul {
  list-style: outside disc;
}

.c11 ol {
  list-style: decimal;
}

.c11 li {
  padding-left: 10px;
}

.c8 {
  font-family: 'Roboto mono',monospace;
  font-weight: 500;
  color: #FBFBFB;
  font-size: 1rem;
  line-height: normal;
  text-transform: uppercase;
}

.c13 {
  position: relative;
  position: relative;
}

.c13::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: currentColor;
}

.c13::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.25s ease-out;
}

.c13:hover::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c13:hover,
.c13:active,
.c13:focus {
  text-decoration: none;
}

.c13:focus-visible {
  outline: #FBFBFB auto 1px;
}

.c4 {
  padding: 0;
  display: flex;
  align-items: center;
  gap: 5px;
  background: transparent;
  outline: none;
  box-shadow: none;
  color: #FBFBFB;
  padding-block: 0;
  padding-inline: 0;
  border: none;
  cursor: pointer;
  position: relative;
}

.c4::after {
  content: '';
  position: absolute;
  width: 100%;
  height: 2px;
  bottom: 0;
  left: 0;
  background-color: currentColor;
}

.c4::after {
  transform: scaleX(0);
  transform-origin: bottom right;
  transition: transform 0.25s ease-out;
}

.c4:hover::after {
  transform: scaleX(1);
  transform-origin: bottom left;
}

.c4:focus-visible {
  outline: #FBFBFB auto 1px;
}

.c0 {
  position: fixed;
  z-index: 1040;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: #000000;
  opacity: 0.7;
}

.c1 {
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  position: fixed;
  z-index: 1040;
  border: 1px solid #FBFBFB;
  border-radius: 2px;
  background-color: #000000;
}

.c3 {
  position: relative;
  padding: 40px;
}

.c5 {
  position: absolute;
  top: 20px;
  right: 20px;
  color: #FBFBFB;
}

.c5:after {
  content: none;
}

.c5:hover:after {
  content: none;
}

.c7 {
  margin-bottom: 40px;
}

.c9 {
  margin: 20px 0 10px;
}

.c10 {
  display: flex;
  flex-wrap: wrap;
  gap: 8px;
}

.c14:after {
  content: none;
}

.c14:hover:after {
  content: none;
}

.c12 {
  list-style: none;
  display: flex;
  gap: 1rem;
}

.c2 {
  border: 1px solid;
  border-image-slice: 1;
  border-image-source: linear-gradient(           to right,           #FFDD83,           #98DFD6         );
}

@media (width >= 768px) {
  .c1 {
    top: 30%;
    left: 50%;
    transform: translate(-51%, -40%);
    width: 100%;
    height: auto;
    max-width: 800px;
    box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
  }
}

<body
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="c0"
  />
  <div
    aria-labelledby="about-modal-label"
    aria-modal="true"
    class="c1 c2"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="c3"
      role="document"
    >
      <button
        class="c4 c5"
      >
        <svg
          height="24"
          width="24"
        >
          close.svg
        </svg>
      </button>
      <section>
        <h4
          class="c6 c7"
          id="about-modal-label"
        >
          About this page
        </h4>
        <div
          class="c8 c9"
        >
          Source code
        </div>
        <div
          class="c10"
        >
          <p
            class="c11"
          >
            What I've used:
          </p>
          <ul
            class="c12"
          >
            <li>
              <a
                aria-label="NextJS"
                class="c13 c14"
                data-testid="plain-anchor"
                href="https://nextjs.org/"
                rel="noreferrer"
                target="_blank"
              >
                <svg
                  height="20"
                  width="20"
                >
                  nextjs.svg
                </svg>
              </a>
            </li>
            <li>
              <a
                aria-label="Styled-components"
                class="c13 c14"
                data-testid="plain-anchor"
                href="https://styled-components.com/"
                rel="noreferrer"
                target="_blank"
              >
                <svg
                  height="20"
                  width="20"
                >
                  styledComponents.svg
                </svg>
              </a>
            </li>
            <li>
              <a
                aria-label="Jest"
                class="c13 c14"
                data-testid="plain-anchor"
                href="https://jestjs.io/"
                rel="noreferrer"
                target="_blank"
              >
                <svg
                  height="20"
                  width="20"
                >
                  jest.svg
                </svg>
              </a>
            </li>
            <li>
              <a
                aria-label="NX"
                class="c13 c14"
                data-testid="plain-anchor"
                href="https://nx.dev/"
                rel="noreferrer"
                target="_blank"
              >
                <svg
                  height="20"
                  width="20"
                >
                  nx.svg
                </svg>
              </a>
            </li>
            <li>
              <a
                aria-label="Yarn"
                class="c13 c14"
                data-testid="plain-anchor"
                href="https://yarnpkg.com/"
                rel="noreferrer"
                target="_blank"
              >
                <svg
                  height="20"
                  width="20"
                >
                  yarn.svg
                </svg>
              </a>
            </li>
          </ul>
        </div>
        <div
          class="c10"
        >
          <p
            class="c11"
          >
            Repository:
          </p>
          <a
            aria-label="Source code repository link"
            class="c13 c14"
            data-testid="plain-anchor"
            href="https://github.com/sauldeleon/portfolio-blog"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              height="20"
              width="20"
            >
              github.svg
            </svg>
          </a>
        </div>
        <div
          class="c8 c9"
        >
          Design
        </div>
        <div
          class="c10"
        >
          <p
            class="c11"
          >
            Designed by: Valentina Florentina Balta-Cojocaru-Stan
          </p>
          <a
            aria-label="Valentina Florentina Balta-Cojocaru-Stan LinkedIn profile link"
            class="c13 c14"
            data-testid="plain-anchor"
            href="https://www.linkedin.com/in/valentina-florentina-balta-cojocaru-stan-83619014a/"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              height="20"
              width="20"
            >
              linkedIn.svg
            </svg>
          </a>
        </div>
        <div
          class="c10"
        >
          <p
            class="c11"
          >
            Design:
          </p>
          <a
            aria-label="Figma design link"
            class="c13 c14"
            data-testid="plain-anchor"
            href="https://www.figma.com/file/ZgfaiU473XMQAWz9kbiMO6/Saul-portfolio-%2B-design-library"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              height="20"
              width="20"
            >
              figma.svg
            </svg>
          </a>
        </div>
      </section>
    </div>
  </div>
</body>
`;
